### PR TITLE
Remove Redundant Function for Extracting Scientific Notation from Strings in DataGen.

### PR DESCRIPTION
**Pull request by younies**

### CODE REVIEW

File: provider/datagen/Cargo.toml [hunk 1]
```diff
@@ -85,6 +85,7 @@ itertools = { version = "0.10", optional = true }
 ndarray = { version = "0.15.5", default-features = false, optional = true }
 num-bigint = { version = "0.4.4", default-features = false, optional = true }
 num-rational = { version = "0.4", default-features = false, optional = true }
+num-traits = { version = "0.2.17", default-features = false}
```
Comment (younies) [inline provider/datagen/Cargo.toml:88]: @robertbastian : if I added `optional = true`, I could not access `use num_traits::One;`. Shall I add a specific configurations?

File: provider/datagen/src/transform/cldr/units/helpers.rs [hunk 2]
```diff
@@ -139,18 +138,18 @@ pub(in crate::provider) fn process_factor(
...
-
 /// Extracts the conversion info from a base unit, factor and offset.
 pub(in crate::provider) fn extract_conversion_info<'data>(
     base_unit: &str,
     factor: &ScientificNumber,
     offset: &ScientificNumber,
     parser: &MeasureUnitParser,
 ) -> Result<ConversionInfo<'data>, DataError> {
+    /// Converts a vector of strings to a vector of string slices.
+    fn to_str_vec(string_vec: &[String]) -> Vec<&str> {
```
Comment (robertbastian) [inline provider/datagen/src/transform/cldr/units/helpers.rs:149]: you don't need this if you change `convert_slices_to_fraction` to accept `&[String]`

File: provider/datagen/Cargo.toml (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/Cargo.toml:88]: I don't see why you wouldn't be able to. What's the error?

Comment (younies) [reply provider/datagen/Cargo.toml:88]: here is the error:

```
error[E0432]: unresolved import `num_traits`
  --> provider/datagen/src/transform/cldr/units/helpers.rs:12:5
   |
12 | use num_traits::One;
   |     ^^^^^^^^^^ use of undeclared crate or module `num_traits`

error[E0599]: no function or associated item named `one` found for struct `IcuRatio` in the current scope
   --> provider/datagen/src/transform/cldr/units/helpers.rs:353:29
    |
353 |         .try_fold(IcuRatio::one(), |result, &num| {
    |                             ^^^ function or associated item not found in `IcuRatio`
```

File: provider/datagen/src/transform/cldr/units/helpers.rs (hunk 2 above)
Comment (younies) [reply provider/datagen/src/transform/cldr/units/helpers.rs]: good idea

File: provider/datagen/Cargo.toml (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/Cargo.toml]: have you added it to the experimental feature?

File: components/experimental/src/units/ratio.rs [hunk 3]
```diff
@@ -54,10 +54,23 @@ pub enum RatioFromStrError {
...
+    /// Retrieves the numerator of the ratio.
+    /// NOTE:
+    ///     The numerator retains the sign of the overall ratio.
+    ///     For instance, in the ratio -2/3, the numerator is -2.
+    pub fn numer(&self) -> &BigInt {
+        &self.0.numer()
+    }
+
+    /// Retrieves the denominator of the ratio.
+    pub fn denom(&self) -> &BigInt {
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:70]: yes, should `IcuRatio` use `BigUint` for the denominator if the sign is always part of the nominator?

[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/ratio.rs:70]: We can make this implementation, but in this case, we will not depend on Ration at all, we will implement our own IcuRatio.

struct IcuRatio {
  num: BigInt,
  den: BigUint,
}

and we will lose a lot of functionalities that associated with `Ratio`.

File: provider/datagen/Cargo.toml (hunk 1 above)
Comment (younies) [reply provider/datagen/Cargo.toml]: yes, it seems I need to added to `experimental_components`

File: components/experimental/src/units/ratio.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:70]: oh, but `BigRational` does not guarantee that `numer()` retains the sign of the overall ratio afaict. You might just want to have a `as_big_rational()` getter that exposes the `BigRational`. Then callers can use `numer()` and `denom()` with `BigRational`'s guarantees.

[the remark this answers is not in the document]
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:70]: This is not guaranteed and could change tomorrow.

[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/ratio.rs:70]: but you are right, let's split this

```
    #[inline]
    fn is_positive(&self) -> bool {
        (self.numer.is_positive() && self.denom.is_positive())
            || (self.numer.is_negative() && self.denom.is_negative())
    }

    #[inline]
    fn is_negative(&self) -> bool {
        (self.numer.is_negative() && self.denom.is_positive())
            || (self.numer.is_positive() && self.denom.is_negative())
    }
```

[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/ratio.rs:67]: Even it is true that `numer` is always having the sing, but it is not stated clearly that will be always the case.

[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/ratio.rs:67]: or they can call `is_negative` and `is_positive` from IcuRatio  too

[the remark this answers is not in the document]
Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:67]: let's not duplicate functionality

**Review by robertbastian — COMMENTED** [review]
.

File: components/experimental/src/units/ratio.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

File: components/experimental/src/units/ratio.rs [hunk 4]
```diff
@@ -432,6 +437,21 @@ impl FromStr for IcuRatio {
 mod tests {
     use super::*;
 
+    #[test]
+    fn test_ratio_always_keep_sign_in_numerator() {
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:441]: remove?

**Review by robertbastian — APPROVED** [review]

File: components/experimental/src/units/ratio.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:441]: sure, in the following PR

Comment (younies) [reply components/experimental/src/units/ratio.rs:441]: https://github.com/unicode-org/icu4x/pull/4835

**Pull request merged**